src/middlewared/middlewared/plugins/directoryservices.py

```diff
@@ -110,6 +176,126 @@ async def nss_info_choices(self, dstype):
 
         return ret
 
+    @private
+    def backup_secrets(self):
+        ha_mode = self.middleware.call_sync('smb.get_smb_ha_mode')
+        netbios_name = self.middleware.call_sync('smb.config')['netbiosname_local']
+        db = self.middleware.call_sync('datastore.query',
+                                       'services.cifs', [],
+                                       {'prefix': 'cifs_srv_', 'get': True})
+
+        if db['secrets'] is None:
+            db_secrets = {}
+        else:
+            db_secrets = json.loads(db['secrets'])
+
+        with DirectorySecrets(logger=self.logger, ha_mode=ha_mode) as s:
+            secrets = s.dump()
+
+        if not secrets:
+            self.logger.warning("Unable to parse secrets")
+            return
+
+        db_secrets.update({netbios_name: secrets})
+        self.middleware.call_sync('datastore.update',
+                                  'services.cifs', 1,
+                                  {'secrets': str(json.dumps(db_secrets))},
+                                  {'prefix': 'cifs_srv_'})
+
+    @private
+    def restore_secrets(self, netbios_name=None):
+        ha_mode = self.middleware.call_sync('smb.get_smb_ha_mode')
+
+        if netbios_name is None:
+            netbios_name = self.middleware.call_sync('smb.config')['netbiosname_local']
+
+        db = self.middleware.call_sync('datastore.query',
+                                       'services.cifs', [],
+                                       {'prefix': 'cifs_srv_', 'get': True})
+
+        db_secrets = json.loads(db['secrets'])
+        server_secrets = db_secrets.get(netbios_name)
+        if server_secrets is None:
+            self.logger.warning("Unable to find stored secrets for [%]. "
+                                "Directory service functionality may be impacted.",
+                                netbios_name)
+            return False
+
+        with DirectorySecrets(logger=self.logger, ha_mode=ha_mode) as s:
+            try:
+                s.restore(server_secrets)
+            except Exception:
+                self.logger.warning("Failed to restore secrets for [%s]. "
+                                    "Directory service functionality may be impacted.",
+                                    netbios_name, exc_info=True)
+                return False
+
+        return True
+
+    @private
+    def secrets_has_domain(self, domain):
+        ha_mode = self.middleware.call_sync('smb.get_smb_ha_mode')
+        with DirectorySecrets(logger=self.logger, ha_mode=ha_mode) as s:
+            rv = s.has_domain(domain)
+
+        return rv
+
+    @private
+    def available_secrets(self):
+        db = self.middleware.call_sync('datastore.query',
+                                       'services.cifs', [],
+                                       {'prefix': 'cifs_srv_', 'get': True})
+
+        db_secrets = json.loads(db['secrets'])
```

Comment:
Are we always going to have valid json values in `db['secrets']` ?

Reply:
We should, but I'll catch JSON decoding errors here.

Reply:
I'll circle back around and fix any further concerns once I have some time to test against HA.